packages/opentelemetry-api/src/diag/logger.ts

```diff
@@ -0,0 +1,158 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { Logger } from '../common/Logger';
+
+/** Defines a type which can be used for as a parameter without breaking backward
+ * compatibility. The {@link Logger} reference will be removed with the removal
+ * of the Logger definition, this can be used as a replacement for functions
+ * that are currently passing Logger references during migration to minimize
+ * breaks that will occur with the removal of the Logger interface.
+ */
+export type OptionalDiagLogger = Logger | DiagLogger | null | undefined;
+
+export type DiagLogFunction = (message: string, ...args: unknown[]) => void;
+
+/** Defines an internal diagnostic logger interface which is used to log internal diagnostic
+ * messages, you can set the default diagnostic logger via the {@link DiagAPI} setLogger function.
+ * API provided implementations include :-
+ * - a No-Op {@link noopDiagLogger}
+ * - a {@link DiagLogLevel} filtering wrapper {@link diagLogLevelFilter}
+ * - a general Console {@link DiagConsoleLogger} version.
+ */
+export interface DiagLogger {
+  /** Log a terminal situation that would cause the API to completely fail to initialize,
+   * if this type of message is logged functionality of the API is not expected to be functional.
+   */
+  terminal: DiagLogFunction;
+
+  /** Log a critical error that NEEDS to be addressed, functionality of the component that emits
+   * this log detail may non-functional. While the overall API may be.
```

Comment:
Should be 
this log detail may **be** non-functional. While the overall API may be.

Trying to portray that rather than throwing and aborting the execution flow, whatever logs this level of message may or may not continue to be functional.. eg. If this occurs during initialization then that component may consider itself as not-initialized and therefore just won't work, while if it occurs during processing an event (as as an external connection) or total failure of any retry logic, it will either just be a temporary situation etc.

Reply:
Reworked a little